clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -103,9 +103,16 @@ public void configure(TransportLayer transportLayer, PrincipalBuilder principalB
             mechanism = (String) this.configs.get(SaslConfigs.SASL_MECHANISM);
             if (mechanism == null)
                 throw new IllegalArgumentException("SASL mechanism not specified");
-            // Since 0.9.0.x servers expect to see GSSAPI packets without mechanism being sent first, send
-            // mechanism to server only for non-GSSAPI.
-            setSaslState(mechanism.equals(SaslConfigs.GSSAPI_MECHANISM) ? SaslState.INITIAL : SaslState.SEND_HANDSHAKE_REQUEST);
+
+            // Since 0.9.0.x brokers expect to see GSSAPI packets without Kafka handshake requests, skip
+            // handshake request phase for inter-broker communication with 0.9.0.x. Since newer clients with
+            // version 0.10 and above do not interoperate with 0.9.0.x servers, we can send handshake
+            // requests for client connections which are not for inter-broker communication.
+            boolean skipHandshakeRequest = configs.containsKey("inter.broker.protocol.version") &&
+                    ((String) configs.get("inter.broker.protocol.version")).startsWith("0.9.0");
+            if (skipHandshakeRequest && !mechanism.equals(SaslConfigs.GSSAPI_MECHANISM))
+                throw new IllegalArgumentException("Only GSSAPI is supported with Kafka 0.9.0.x");
+            setSaslState(skipHandshakeRequest ? SaslState.INITIAL : SaslState.SEND_HANDSHAKE_REQUEST);
```

Comment:
Hmm, we want to check inter.broker.protocol.version >= 0.10.0. This is easier if we can use the case object in core. Since we only need to use the old protocol when SaslClientAuthethicator is used at the broker side. Perhaps, we can check  inter.broker.protocol.version in the broker code and pass a flag into inter.broker.protocol.version. The places where we use SaslClientAuthethicator are in ReplicaFetcherThread, ControllerChannelManager, and KafkaServer (for controlled shutdown). When used in clients (producer/consumer), SaslClientAuthethicator will always use the new protocol.